crypto/src/proteus.rs

```diff
@@ -462,10 +470,36 @@ impl ProteusCentral {
         &mut self,
         session_id: &str,
         key: &[u8],
-    ) -> CryptoResult<GroupStoreValue<ProteusConversationSession>> {
-        let prekey = PreKeyBundle::deserialise(key).map_err(ProteusError::from)?;
-        let proteus_session =
-            Session::init_from_prekey(self.proteus_identity.clone(), prekey).map_err(ProteusError::from)?;
+    ) -> Result<GroupStoreValue<ProteusConversationSession>> {
+        let prekey = PreKeyBundle::deserialise(key).map_err(ProteusError::wrap("deserializing prekey bundle"))?;
+        // Note on the `::<Error>` turbofish below:
+        //
+        // `init_from_prekey` returns an error type which is parametric over some wrapped `E`,
+        // because one variant (not relevant to this particular operation) wraps an error type based
+        // on a parameter of a different function entirely.
+        //
+        // Rust complains here, because it can't figure out what type that `E` should be. After all, it's
+        // not inferrable from this function call! It is also entirely irrelevant in this case.
+        //
+        // We can derive two general rules about error-handling in Rust from this example:
+        //
+        // 1. It's better to make smaller error types where possible, encapsulating fallible operations
+        //    with their own error variants, and then wrapping those errors where required, as opposed to
+        //    creating giant catch-all errors. Doing so also has knock-on benefits with regard to tracing
+        //    the precise origin of the error.
+        // 2. One should never make an error wrapper parametric. If you need to wrap an unknown error,
+        //    it's always better to wrap a `Box<dyn std::error::Error>` than to make your error type parametric.
+        //    The allocation cost of creating the `Box` is utterly trivial in an error-handling path, and
+        //    it avoids parametric virality. (`init_from_prekey` is itself only generic because it returns
+        //    this error type with a type-parametric variant, which the function never returns.)
+        //
+        // In this case, we could in principle use the unit type, because we never construct the error variant
+        // in question. But of course, that doesn't implement `Display`, which is required for deriving the
+        // wrapper's `Error` implementation.
+        //
+        // So instead we use our own error type, which is recursive and silly, but gets the job done.
+        let proteus_session = Session::init_from_prekey::<Error>(self.proteus_identity.clone(), prekey)
+            .map_err(KeystoreError::wrap("initializing session from prekey"))?;
```

Comment:
Is this `KeystoreError::wrap` a typo, i.e. shouldn't it be `ProteusError::wrap`?

Reply:
Oops, you're right: https://github.com/wireapp/core-crypto/pull/791/commits/80ae45aa3fc87e88c9bad593782699737b648538